File: src/main/java/com/arnav/tutionSAAS/controller/ParentController.java
package com.arnav.tutionSAAS.controller;

import com.arnav.tutionSAAS.dto.LinkedStudentResponse;
import com.arnav.tutionSAAS.dto.ParentLinkRequest;
import com.arnav.tutionSAAS.entity.User;
import com.arnav.tutionSAAS.service.AttendanceService;
import com.arnav.tutionSAAS.service.ParentService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/parent")
@PreAuthorize("hasRole('PARENT')")
public class ParentController {

    @Autowired private ParentService parentService;

    /**
     * Parent uses the 6-char code from their child to link accounts.
     */
    @PostMapping("/link")
    public ResponseEntity<Void> linkToStudent(
            @AuthenticationPrincipal Jwt jwt,
            @RequestBody ParentLinkRequest request) {
        parentService.linkToStudent(jwt.getSubject(), request.getLinkCode());
        return ResponseEntity.noContent().build();
    }

    /**
     * Returns all students linked to this parent with fee and blocked status.
     */
    @GetMapping("/students")
    public ResponseEntity<List<LinkedStudentResponse>> getLinkedStudents(@AuthenticationPrincipal Jwt jwt) {
        return ResponseEntity.ok(parentService.getLinkedStudents(jwt.getSubject()));
    }

    /**
     * Returns attendance summary for a linked student in a specific batch.
     */
    @GetMapping("/students/{studentId}/attendance")
    public ResponseEntity<AttendanceService.AttendanceSummary> getStudentAttendance(
            @AuthenticationPrincipal Jwt jwt,
            @PathVariable Long studentId,
            @RequestParam Long batchId) {
        return ResponseEntity.ok(parentService.getStudentAttendance(jwt.getSubject(), studentId, batchId));
    }

    /**
     * Returns fee status for a linked student.
     */
    @GetMapping("/students/{studentId}/fees")
    public ResponseEntity<Map<String, Object>> getStudentFeeStatus(
            @AuthenticationPrincipal Jwt jwt,
            @PathVariable Long studentId) {
        return ResponseEntity.ok(parentService.getStudentFeeStatus(jwt.getSubject(), studentId));
    }
    
    @Autowired private com.arnav.tutionSAAS.service.RazorpayService razorpayService;
    @Autowired private com.arnav.tutionSAAS.repository.UserRepo userRepo;
    @Autowired private com.arnav.tutionSAAS.repository.BatchRepo batchRepo;

    @PostMapping("/students/{studentId}/fees/order")
    public ResponseEntity<Map<String, Object>> createOrder(
            @AuthenticationPrincipal Jwt jwt,
            @PathVariable Long studentId) {
        
        parentService.getStudentFeeStatus(jwt.getSubject(), studentId); // Asserts link
        
        User student = userRepo.findById(studentId)
                .orElseThrow(() -> new RuntimeException("Student not found"));

        List<com.arnav.tutionSAAS.entity.Batch> batches = batchRepo.findByStudents_Id(studentId);
        double totalFee = batches.stream().mapToDouble(com.arnav.tutionSAAS.entity.Batch::getMonthlyFee).sum();
        
        if (totalFee <= 0) {
            return ResponseEntity.badRequest().body(Map.of("error", "No fees due"));
        }

        String orderId = razorpayService.createOrder(totalFee, student);

        return ResponseEntity.ok(Map.of(
            "orderId", orderId,
            "amount", totalFee,
            "currency", "INR"
        ));
    }

    @PostMapping("/students/{studentId}/fees/verify")
    public ResponseEntity<Map<String, Object>> verifyPayment(
            @AuthenticationPrincipal Jwt jwt,
            @PathVariable Long studentId,
            @RequestBody Map<String, String> payload) {
        
        parentService.getStudentFeeStatus(jwt.getSubject(), studentId); // Asserts link
        
        String razorpayOrderId = payload.get("razorpay_order_id");
        String razorpayPaymentId = payload.get("razorpay_payment_id");
        String razorpaySignature = payload.get("razorpay_signature");

        boolean isValid = razorpayService.verifyPayment(razorpayOrderId, razorpayPaymentId, razorpaySignature);

        if (isValid) {
            return ResponseEntity.ok(Map.of(
                "success", true,
                "message", "Payment verified successfully"
            ));
        } else {
            return ResponseEntity.badRequest().body(Map.of(
                "success", false,
                "message", "Payment signature verification failed"
            ));
        }
    }
    
    @GetMapping("/students/{studentId}/batches")
    public ResponseEntity<List<com.arnav.tutionSAAS.dto.BatchResponse>> getStudentBatches(
            @AuthenticationPrincipal Jwt jwt,
            @PathVariable Long studentId) {
        return ResponseEntity.ok(parentService.getStudentBatches(jwt.getSubject(), studentId));
    }

    @GetMapping("/students/{studentId}/sessions")
    public ResponseEntity<List<com.arnav.tutionSAAS.dto.ClassSessionResponse>> getStudentSessions(
            @AuthenticationPrincipal Jwt jwt,
            @PathVariable Long studentId) {
        return ResponseEntity.ok(parentService.getStudentSessions(jwt.getSubject(), studentId));
    }
}

File: backend-nest/src/parent/parent.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Get,
  Param,
  ParseIntPipe,
  Post,
  Query,
  UseGuards,
} from '@nestjs/common';
import * as crypto from 'crypto';
import Razorpay from 'razorpay';
import { Role } from '@prisma/client';
import { ConfigService } from '@nestjs/config';
import { ClerkAuthGuard } from '../common/guards/clerk-auth.guard';
import { RolesGuard } from '../common/guards/roles.guard';
import { Roles } from '../common/decorators/roles.decorator';
import { ClerkId } from '../common/decorators/current-user.decorator';
import { ParentService } from './parent.service';
import { PrismaService } from '../prisma/prisma.service';

@Controller('api/parent')
@UseGuards(ClerkAuthGuard, RolesGuard)
@Roles(Role.PARENT)
export class ParentController {
  private razorpay: any;

  constructor(
    private parentService: ParentService,
    private prisma: PrismaService,
    private config: ConfigService,
  ) {
    this.razorpay = new Razorpay({
      key_id: config.get('RAZORPAY_KEY_ID'),
      key_secret: config.get('RAZORPAY_KEY_SECRET'),
    });
  }

  @Post('link')
  async linkToStudent(
    @ClerkId() clerkId: string,
    @Body() body: { linkCode: string },
  ) {
    await this.parentService.linkToStudent(clerkId, body.linkCode);
    return null;
  }

  @Get('students')
  getLinkedStudents(@ClerkId() clerkId: string) {
    return this.parentService.getLinkedStudents(clerkId);
  }

  @Get('students/:studentId/attendance')
  getStudentAttendance(
    @ClerkId() clerkId: string,
    @Param('studentId', ParseIntPipe) studentId: number,
    @Query('batchId', ParseIntPipe) batchId: number,
  ) {
    return this.parentService.getStudentAttendance(clerkId, studentId, batchId);
  }

  @Get('students/:studentId/fees')
  getStudentFeeStatus(
    @ClerkId() clerkId: string,
    @Param('studentId', ParseIntPipe) studentId: number,
  ) {
    return this.parentService.getStudentFeeStatus(clerkId, studentId);
  }

  @Get('students/:studentId/batches')
  getStudentBatches(
    @ClerkId() clerkId: string,
    @Param('studentId', ParseIntPipe) studentId: number,
  ) {
    return this.parentService.getStudentBatches(clerkId, studentId);
  }

  @Get('students/:studentId/sessions')
  getStudentSessions(
    @ClerkId() clerkId: string,
    @Param('studentId', ParseIntPipe) studentId: number,
  ) {
    return this.parentService.getStudentSessions(clerkId, studentId);
  }

  @Post('students/:studentId/fees/order')
  async createOrder(
    @ClerkId() clerkId: string,
    @Param('studentId', ParseIntPipe) studentId: number,
  ) {
    await this.parentService.getStudentFeeStatus(clerkId, studentId); // asserts link

    const student = await this.prisma.user.findUnique({ where: { id: studentId } });
    if (!student) throw new BadRequestException('Student not found');

    const batches = await this.prisma.batch.findMany({
      where: { students: { some: { id: studentId } } },
    });
    const totalFee = batches.reduce((sum, b) => sum + b.monthlyFee, 0);
    if (totalFee <= 0) throw new BadRequestException('No fees due');

    const order = await this.razorpay.orders.create({
      amount: Math.round(totalFee * 100),
      currency: 'INR',
      receipt: `txn_${Date.now()}`,
    });

    await this.prisma.paymentHistory.create({
      data: {
        studentId,
        razorpayOrderId: order.id,
        amount: totalFee,
        status: 'CREATED',
      },
    });

    return { orderId: order.id, amount: totalFee, currency: 'INR' };
  }

  @Post('students/:studentId/fees/verify')
  async verifyPayment(
    @ClerkId() clerkId: string,
    @Param('studentId', ParseIntPipe) studentId: number,
    @Body() body: any,
  ) {
    await this.parentService.getStudentFeeStatus(clerkId, studentId); // asserts link

    const { razorpay_order_id, razorpay_payment_id, razorpay_signature } = body;
    const keySecret = this.config.get('RAZORPAY_KEY_SECRET');

    const generated = crypto
      .createHmac('sha256', keySecret)
      .update(`${razorpay_order_id}|${razorpay_payment_id}`)
      .digest('hex');

    const isValid = generated === razorpay_signature;

    const payment = await this.prisma.paymentHistory.findUnique({
      where: { razorpayOrderId: razorpay_order_id },
    });

    if (payment) {
      await this.prisma.paymentHistory.update({
        where: { id: payment.id },
        data: {
          razorpayPaymentId: razorpay_payment_id,
          status: isValid ? 'SUCCESS' : 'FAILED',
        },
      });
    }

    if (isValid) {
      await this.prisma.studentProfile.update({
        where: { id: studentId },
        data: { feesPaidForCurrentMonth: true },
      });
      return { success: true, message: 'Payment verified successfully' };
    }
    return { success: false, message: 'Payment signature verification failed' };
  }
}
